### PR TITLE
M-03 config validation

### PR DESCRIPTION
**Pull request by bachstatter**
- Allow withdrawals even if the collateral is configured to maxAllowable: 0
- Deposits should still respect maxAllowable
- Keep track of active market ids, and add `getActiveMarketIds` 
- Add a method to get total collateral amount across all markets
- Make sure we dont remove collateral that have money in the system, but still allow maxAllowable: 0

### CODE REVIEW

File: test/integration/modules/MarginModule.test.ts [hunk 1]
```diff
@@ -1486,31 +1522,55 @@ describe('MarginModule', async () => {
...
       await PerpMarketProxy.connect(from).setCollateralConfiguration(synthMarketIds, maxAllowables);
 
-      // Depositing should cause a failure.
-      const { market, trader, collateral, collateralDepositAmount } = await mintAndApproveWithTrader(bs, genTrader(bs));
+      const configuredCollaterals = await PerpMarketProxy.connect(from).getConfiguredCollaterals();
+      assertBn.isZero(configuredCollaterals[0].maxAllowable);
+      assertBn.isZero(configuredCollaterals[1].maxAllowable);
+    });
+
+    it('should not allow removal of collateral with amounts in the system', async () => {
```
Comment (davidvuong) [inline test/integration/modules/MarginModule.test.ts:1542]: Minor: Can we use the `should revert when xxx` structure for tests that assert for reverts

File: test/integration/modules/MarginModule.test.ts [hunk 2]
```diff
@@ -555,6 +555,42 @@ describe('MarginModule', async () => {
         await assertEvent(receipt, `MarginWithdraw(${marginWithdrawEventProperties})`, PerpMarketProxy);
       });
 
+      it('should allow withdraw of collateral when collateral maxAllowable is 0', async () => {
```
Comment (davidvuong) [inline test/integration/modules/MarginModule.test.ts:558]: Can we also add a test for withdrawAllMargin for this too?

File: contracts/storage/GlobalPerpMarket.sol [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    function getTotalCollateralAmounts(
+        GlobalPerpMarket.Data storage self,
+        uint128 synthMarketId
+    ) internal view returns (uint256) {
+        uint128[] memory activeMarketIds = self.activeMarketIds;
+
+        uint256 activeMarketIdsLength = activeMarketIds.length;
+        uint256 totalCollateralAmount;
+        // Accumulate collateral amounts for active markets
+        for (uint256 i = 0; i < activeMarketIdsLength; i++) {
```
Comment (davidvuong) [inline contracts/storage/GlobalPerpMarket.sol:30]: We're still on 0.8.19 so `unchecked { ++i  }` for the gas savings.

File: contracts/storage/GlobalPerpMarket.sol [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    function load() internal pure returns (GlobalPerpMarket.Data storage d) {
+        bytes32 s = SLOT_NAME;
+
+        assembly {
+            d.slot := s
+        }
+    }
+
+    function getTotalCollateralAmounts(
```
Comment (davidvuong) [inline contracts/storage/GlobalPerpMarket.sol:21]: Add `@dev` documentation here

File: contracts/modules/PerpMarketFactoryModule.sol [hunk 5]
```diff
@@ -131,6 +135,10 @@ contract PerpMarketFactoryModule is IPerpMarketFactoryModule {
     }
 
     // --- Views --- //
+    function getActiveMarketIds() external view returns (uint128[] memory) {
```
Comment (davidvuong) [inline contracts/modules/PerpMarketFactoryModule.sol:138]: Update `IPerpMarketFactoryModule` to include this, add `@notice` docs here

File: contracts/modules/PerpMarketFactoryModule.sol [hunk 6]
```diff
@@ -131,6 +135,10 @@ contract PerpMarketFactoryModule is IPerpMarketFactoryModule {
...
+        GlobalPerpMarket.Data storage global = GlobalPerpMarket.load();
```
Comment (davidvuong) [inline contracts/modules/PerpMarketFactoryModule.sol:139]: Minor: Can just `return GlobalPerpMarket.load().activeMarketIds`;

File: contracts/modules/PerpMarketFactoryModule.sol [hunk 7]
```diff
@@ -73,9 +74,12 @@ contract PerpMarketFactoryModule is IPerpMarketFactoryModule {
         OwnableStorage.onlyOwner();
 
         PerpMarketConfiguration.GlobalData storage globalConfig = PerpMarketConfiguration.load();
+        GlobalPerpMarket.Data storage global = GlobalPerpMarket.load();
+
         uint128 id = globalConfig.synthetix.registerMarket(address(this));
 
         PerpMarket.create(id, data.name);
+        global.activeMarketIds.push(id);
```
Comment (davidvuong) [inline contracts/modules/PerpMarketFactoryModule.sol:82]: Minor: Can just `GlobalPerpMarket.load().activeMarketIds.push(id)`

File: contracts/modules/PerpMarketFactoryModule.sol (hunk 5 above)
Comment (bachstatter) [reply contracts/modules/PerpMarketFactoryModule.sol:138]: 👍  Added interface too

**Comment by bachstatter** [comment]
Comments addressed, also two changes discussed on discord.
Instead of creating a new `GlobalPerpMarket`, add `GlobalData` for `PerpMarket`
Replace `getTotalCollateralAmounts` with `isCollateralDeposited`

**Review by davidvuong — APPROVED** [review]
LGTM!

File: contracts/modules/MarginModule.sol [hunk 8]
```diff
@@ -238,6 +239,25 @@ contract MarginModule is IMarginModule {
...
     }
 
+    function isCollateralDeposited(uint128 synthMarketId) internal view returns (bool) {
+        PerpMarket.GlobalData storage globalPerpMarket = PerpMarket.load();
+
+        uint128[] memory activeMarketIds = globalPerpMarket.activeMarketIds;
+        uint256 activeMarketIdsLength = activeMarketIds.length;
+        // Accumulate collateral amounts for active markets
+        for (uint256 i = 0; i < activeMarketIdsLength; ) {
+            uint128 marketId = activeMarketIds[i];
```
Comment (davidvuong) [inline contracts/modules/MarginModule.sol:249]: Minor: You can avoid creating a var on every iteration by just passing it into `PerpMarket.load`

Comment (bachstatter) [reply contracts/modules/MarginModule.sol]: yep done

**Pull request merged**